pkg/profiler/profiler.go

```diff
@@ -334,77 +294,114 @@ func (p *CgroupProfiler) Run(ctx context.Context) error {
-
-		// We sample at 100Hz, which is every 10 Million nanoseconds.
-		PeriodType: &profile.ValueType{
-			Type: "cpu",
-			Unit: "nanoseconds",
-		},
-		Period: 10000000,
-	}
-
-	mapping := maps.NewMapping(p.pidMappingFileCache)
-	kernelMapping := &profile.Mapping{
-		// TODO(kakkoyun): Check if this conflicts with https://github.com/google/pprof/pull/675/files
-		File: "[kernel.kallsyms]",
-	}
-	kernelFunctions := map[uint64]*profile.Function{}
-	userFunctions := map[[2]uint64]*profile.Function{}
+	defer func() {
+		if err := recover(); err != nil {
+			level.Error(p.logger).Log("msg", "profile loop recovered from panic", "err", err)
+			debug.PrintStack()
+		}
+	}()
 
-	// 2 uint64 1 for PID and 1 for Addr
-	locations := []*profile.Location{}
-	kernelLocations := []*profile.Location{}
-	kernelAddresses := map[uint64]struct{}{}
-	locationIndices := map[[2]uint64]int{}
-	samples := map[[doubleStackDepth]uint64]*profile.Sample{}
+	var (
+		prof = &profile.Profile{
+			SampleType: []*profile.ValueType{{
+				Type: "samples",
+				Unit: "count",
+			}},
+			TimeNanos:     captureTime.UnixNano(),
+			DurationNanos: int64(p.profilingDuration),
 
+			// We sample at 100Hz, which is every 10 Million nanoseconds.
+			PeriodType: &profile.ValueType{
+				Type: "cpu",
+				Unit: "nanoseconds",
+			},
+			Period: 10000000,
+		}
+		mapping       = maps.NewMapping(p.pidMappingFileCache)
+		kernelMapping = &profile.Mapping{
+			// TODO(kakkoyun): Check if this conflicts with https://github.com/google/pprof/pull/675/files
+			File: "[kernel.kallsyms]",
+		}
+		kernelFunctions = map[uint64]*profile.Function{}
+		userFunctions   = map[[2]uint64]*profile.Function{} // 2 uint64 1 for PID and 1 for Addr
+		locations       = []*profile.Location{}
+		kernelLocations = []*profile.Location{}
+		kernelAddresses = map[uint64]struct{}{}
+		locationIndices = map[[2]uint64]int{}
+		samples         = map[[doubleStackDepth]uint64]*profile.Sample{}
+		byteOrder       = byteorder.GetHostByteOrder()
+
+		// Variables needed for eBPF map batch iteration.
+		countKeysPtr = unsafe.Pointer(&p.countKeys[0])
+		nextCountKey = uintptr(1)
+	)
+
+	// Reset count keys before collecting new traces from the kernel.
+	memsetCountKeys(p.countKeys, stackCountKey{})
+
+	// TODO(kakkoyun): Is there a neat way to find out the number of entries in the map?
+	batchSize := 0
 	it := p.bpfMaps.counts.Iterator()
-	byteOrder := byteorder.GetHostByteOrder()
-
-	// TODO(brancz): Use libbpf batch functions.
 	for it.Next() {
-		// This byte slice is only valid for this iteration, so it must be
-		// copied if we want to do anything with it outside of this loop.
-		keyBytes := it.Key()
+		batchSize++
+	}
+	if err := it.Err(); err != nil {
+		return fmt.Errorf("iterate over counts map: %w", err)
+	}
 
-		r := bytes.NewBuffer(keyBytes)
+	if batchSize == 0 {
+		return nil
+	}
+	level.Debug(p.logger).Log("msg", "fetching stack trace counts in batch", "batchSize", batchSize)
+	time.Sleep(1 * time.Second)
```

Comment:
@brancz :scream:

Reply:
So profiler needs to wait until the BPF program gives the information about stack trace counts. Is that correct? In that case, do you think we can use something like SECCOMP_RET_USER_NOTIF (https://lwn.net/Articles/756233/)?

Reply:
This wait should've happened. We still don't know the reason why it works with it. I'm still investigating. Please see the linked issue. 

And thanks for the tip that might help us here, I'll check.